File: examples/php/Case03Test.php
<?php
declare(strict_types=1);
require_once __DIR__ . '/Case03.php';

$weak = new Case03OrderDb();
commitBusinessStateOnly($weak); // Process stops before publish is invoked.
if (!$weak->confirmed || $weak->outboxPending) throw new RuntimeException('weak gap missing');
$rolledBack = new Case03OrderDb();
try { confirmWithOutbox($rolledBack, true); throw new RuntimeException('expected rollback'); }
catch (RuntimeException $error) {
    if ($rolledBack->confirmed || $rolledBack->outboxPending) throw new RuntimeException('partial commit');
}
$repaired = new Case03OrderDb();
confirmWithOutbox($repaired, false);
if (!$repaired->confirmed || !$repaired->outboxPending) throw new RuntimeException('intent missing');
$broker = new class implements Case03Broker {
    public array $events = [];
    public function publish(string $event): void { $this->events[] = $event; }
};
relayPublishPending($repaired, $broker);
relayPublishPending($repaired, $broker); // Crash before marking SENT.
if ($broker->events !== ['E', 'E']) throw new RuntimeException('duplicate boundary missing');
markPublicationComplete($repaired);
relayPublishPending($repaired, $broker);
if (count($broker->events) !== 2) throw new RuntimeException('published after sent');

File: examples/java/Case03Test.java
import java.util.ArrayList;
import java.util.List;

final class Case03Test {
    public static void main(String[] args) {
        Case03.OrderDb weak = new Case03.OrderDb();
        Case03.commitBusinessStateOnly(weak); // Process stops before publish is invoked.
        if (!weak.confirmed || weak.outboxPending) throw new AssertionError("weak gap missing");
        Case03.OrderDb rolledBack = new Case03.OrderDb();
        try {
            Case03.confirmWithOutbox(rolledBack, true);
            throw new AssertionError("expected rollback");
        } catch (IllegalStateException expected) {
            if (rolledBack.confirmed || rolledBack.outboxPending) throw new AssertionError("partial commit");
        }
        Case03.OrderDb repaired = new Case03.OrderDb();
        Case03.confirmWithOutbox(repaired, false);
        if (!repaired.confirmed || !repaired.outboxPending) throw new AssertionError("intent missing");
        List<String> events = new ArrayList<>();
        Case03.relayPublishPending(repaired, events::add);
        Case03.relayPublishPending(repaired, events::add); // Crash before marking SENT.
        if (events.size() != 2) throw new AssertionError("duplicate boundary missing");
        Case03.markPublicationComplete(repaired);
        Case03.relayPublishPending(repaired, events::add);
        if (events.size() != 2) throw new AssertionError("published after sent");
    }
}
